### PR TITLE
Add CatBoost support

### PR DESCRIPTION
**Pull request by harshitsinghai77**
## Description
Adding support for CatBoost

## Motivation and Context

Add support for serving CatBoost models in BentoML. CatBoost is a high-performance open source library for gradient boosting on decision trees https://catboost.ai/

https://github.com/bentoml/BentoML/issues/1327

## How Has This Been Tested?
Added integration tests

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature and improvements (non-breaking change which adds/improves functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Code Refactoring (internal change which is not user facing)
- [ ] Documentation
- [ ] Test, CI, or build

## Component(s) if applicable
- [ ] BentoService (service definition, dependency management, API input/output adapters)
- [x] Model Artifact (model serialization, multi-framework support)
- [ ] Model Server (mico-batching, dockerisation, logging, OpenAPI, instruments)
- [ ] YataiService gRPC server (model registry, cloud deployment automation)
- [ ] YataiService web server (nodejs HTTP server and web UI)
- [ ] Internal (BentoML's own configuration, logging, utility, exception handling)
- [ ] BentoML CLI

## Checklist:

- [x] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [x] My change reduces project test coverage and requires unit tests to be added
- [x] I have added unit tests covering my code change
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly

### CODE REVIEW

**Comment by harshitsinghai77** [comment]
I need some help with `[BentoML-CI / catboost_integration_tests (pull_request)](https://github.com/bentoml/BentoML/pull/1539/checks?check_run_id=2204893181)`   

`/home/runner/work/_temp/025bb526-8a7e-479e-acf3-3640bc024f03.sh: line 1: ./ci/catboost_integration_tests.sh: Permission denied`

File: README.md [hunk 1]
```diff
@@ -1,24 +1,22 @@
 [<img src="https://raw.githubusercontent.com/bentoml/BentoML/master/docs/source/_static/img/bentoml-readme-header.jpeg" width="600px" margin-left="-5px">](https://github.com/bentoml/BentoML)
 
-## Model Serving Made Easy  [![Tweet](https://img.shields.io/twitter/url/http/shields.io.svg?style=social)](https://twitter.com/intent/tweet?text=BentoML:%20Machine%20Learning%20Model%20Serving%20Made%20Easy%20&url=https://github.com/bentoml/BentoML&via=bentomlai&hashtags=mlops,modelserving,ML,AI,machinelearning,bentoml)
+## Model Serving Made Easy [![Tweet](https://img.shields.io/twitter/url/http/shields.io.svg?style=social)](https://twitter.com/intent/tweet?text=BentoML:%20Machine%20Learning%20Model%20Serving%20Made%20Easy%20&url=https://github.com/bentoml/BentoML&via=bentomlai&hashtags=mlops,modelserving,ML,AI,machinelearning,bentoml)
```
Comment (parano) [inline README.md:3]: Looks like you will need to resolve conflicts with master branch here

Comment (parano) [reply README.md:3]: Or did you apply some kind of auto-formatting for the .md files?  could you tell me what the tool is and maybe we can add it to `make format` script and include these type of changes in a separate PR?

File: docs/source/frameworks.rst [hunk 2]
```diff
@@ -115,6 +115,17 @@ Example Projects:
 .. autoclass:: bentoml.frameworks.lightgbm.LightGBMModelArtifact
 
 
+========
+CatBoost
+========
+
+Example Projects:
+
+* Titanic Survival Prediction -  `Google Colab <https://colab.research.google.com/github/bentoml/gallery/blob/master/lightbgm/titanic-survival-prediction/lightbgm-titanic-survival-prediction.ipynb>`__ / `Notebook Source <https://github.com/bentoml/gallery/blob/master/lightbgm/titanic-survival-prediction/lightbgm-titanic-survival-prediction.ipynb>`__
```
Comment (parano) [inline docs/source/frameworks.rst:124]: Do you plan to add an example notebook for CatBoost in the `bentoml/gallery`  repo?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -2,14 +2,13 @@ name: BentoML-CI
 
 on:
   push:
-    branches: [ master ]
+    branches: [master]
```
Comment (parano) [inline .github/workflows/ci.yml:5]: the same question for this file, did you apply an auto-formatting program for this yml file? and could you make it a separate PR?

File: bentoml/frameworks/catboost.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+class CatBoostModelArtifact(BentoServiceArtifact):
+    """
+    Abstraction for saving/loading CatBoost models.
+
+    Args:
+        name (string): name of the artifact
+        model_type (string): Type of the model - 'classifier', 'regressor', 'catboost'
+        model_extension (string): Extension name for saved CatBoost model. Default - "cbm"
+        model_export_parameters (dict): Additional format-dependent parameters. Default - None
+        model_pool: The dataset previously used for training. See catboost.Pool for more details. Default - None.
```
Comment (yubozhao) [inline bentoml/frameworks/catboost.py:17]: Would the users know ahead of time those two options before training the CatBoost model?

If these two options are changed with newly trained model, I think they would be better fit with the `pack` function.
```
svc.pack('model' CatBoost_Model, model_export=parameters={...}, model_pool={})
```

**Comment by parano** [comment]
Hey @harshitsinghai77, the issue related to artifact init parameters has been fixed in the latest release, which should unblock the issue you were running into with CatBoost artifact class. Let me know if you'd still like to continue to work on this PR.

**Comment by harshit-deepsource** [comment]
Hey. @parano 

I'm busy at the moment. I'd like to work on this PR. Just give me 2 weeks if possible. Thanks.

**Comment by harshitsinghai77** [comment]
Hey. @parano

I'm busy at the moment. I'd like to work on this PR. Just give me 2 weeks if possible. Thanks.

**Comment by aarnphm** [comment]
I will also take over this for the upcoming 1.0 releases

**Comment by aarnphm** [comment]
Done with implementation for 1.0. I will close this for now.

**Comment by harshitsinghai77** [comment]
@aarnphm are the issues fixed? I was going to work on the PR, but if it is fixed then I'll work on some other issue.

**Comment by parano** [comment]
@harshitsinghai77 yes it is done in the 1.0 branch

**Comment by aarnphm** [comment]
I havent merged the PR yet until I finished with all other library
https://github.com/aarnphm/BentoML/blob/framework-redesign/bentoml/catboost.py

**Comment by harshitsinghai77** [comment]
Got it thanks. The changes look good. Let me know if you need any help doing this for other libraries.
